refactor(compiler): Add support for i18n expressions

mmalerba opened the pull request:
Adds support for i18n on templates and i18n expressions. See individual commit messages for more detail.

mmalerba committed:
refactor(compiler): Add support for i18n expressions

Adds support for i18n expressions in i18n messages, and allows i18n
messages on templates.

Co-authored-by: Alex Rickabaugh <alxhub@users.noreply.github.com>
Co-authored-by: Dylan Hunn <dylhunn@users.noreply.github.com>

mmalerba committed:
refactor(compiler): Allow collapsing i18n start/end ops that contain a pipe

Moves the empty element phase earlier, to before pipe creation. This
ensures that adjacent i18nStart/i18nEnd ops will be collapsed into a
isingle i18n op, rather than remaining uncollapsed if a pipe is inserted
between them.

mmalerba committed:
refactor(compiler): Generate correct advance instructions for i18n

Matches the behavior of `TemplateDefinitionBuilder`, advancing to the
last element in the i18n block before evaluating i18n expressions.

mmalerba committed:
refactor(compiler): Add support for applying i18n expressions

Adds support for the i18nApply operation to apply i18n expressions.

mmalerba committed:
refactor(compiler): Fill in i18n expression placeholders

Fills in the correct values for expression placeholders in the i18n
messages.

mmalerba committed:
refactor(compiler): Only pass relevant i18n data to IR

Refactors the i18n handling to only pass the relevant information from
the i18n AST through to the IR, instead of passing the entire
I18nMetadata.

mmalerba committed:
refactor(compiler): Use a real map for i18n params

Use an actual JS `Map` rather than an object to represent the i18n
params.

dylhunn commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -142,22 +148,35 @@ function ingestElement(unit: ViewCompilationUnit, element: t.Element): void {
+  // If there is an i18n message associated with this element, insert i18n start and end ops.
+  if (element.i18n instanceof i18n.Message) {
+    const i18nBlockId = unit.job.allocateXrefId();
+    ir.OpList.insertAfter<ir.CreateOp>(ir.createI18nStartOp(i18nBlockId, element.i18n), startOp);
```
I think these probably should have source maps? Not necessary in this PR of course, but we'll need to figure this out from the tests

mmalerba committed:
refactor(compiler): Refactor handling of pipes in i18n blocks

Rather than rely on the empty element collapsing phase to run first, add
logic to the empty element phase to ignore pipes when deciding whether
to collapse an element.

mmalerba committed:
refactor(compiler): Fix a bug in OpList.replaceWithMany

Fix a bug where replaceWithMany was incorrectly setting the previous
item

mmalerba committed:
refactor(compiler): Enable some more passing i18n tests

Enables a handful of additional i18n tests that pass with the changes
made so far.

mmalerba committed:
refactor(compiler): Additional fixes for pipes in i18n

Adds a new test to verify pipe behavior in i18n blocks, and makes
serveral fixes to allow the test to pass.

mmalerba commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js:
```diff
@@ -0,0 +1,31 @@
+consts: function() {
+  __i18nMsg__('{$interpolation} and {$interpolation_1}', [['interpolation', String.raw`\uFFFD0\uFFFD`], ['interpolation_1', String.raw`\uFFFD1\uFFFD`]], {original_code: {'interpolation': '{{ valueA | pipeA }}', 'interpolation_1': '{{ valueB | pipeB }}'}}, {})
+  __i18nMsg__('{$startTagSpan}{$interpolation}{$closeTagSpan} and {$interpolation_1} {$startTagSpan}and {$interpolation_2}{$closeTagSpan}', [['closeTagSpan', String.raw`[\uFFFD/#6\uFFFD|\uFFFD/#9\uFFFD]`], ['interpolation', String.raw`\uFFFD0\uFFFD`], ['interpolation_1', String.raw`\uFFFD1\uFFFD`], ['interpolation_2', String.raw`\uFFFD2\uFFFD`], ['startTagSpan', String.raw`[\uFFFD#6\uFFFD|\uFFFD# ...
+  …
```
Note: there is an `i18npostprocess` instruction that is supposed to go here, but I haven't implemented support for it yet and this PR was getting big already, so I've ignored it for now

mmalerba commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js:
```diff
@@ -0,0 +1,31 @@
+  }
+  if (rf & 2) {
+    $r3$.ɵɵadvance(3);
+    $r3$.ɵɵi18nExp($r3$.ɵɵpipeBind1(2, …, ctx.valueA))($r3$.ɵɵpipeBind1(3, …, ctx.valueB));
```
Note: the variable offsets for the pipes were incorrect. I'm not sure if this is a bug or not, but its something I've seen in other tests as well. I ignored it for now to focus on making sure we have the right instructions in the right order

mmalerba committed:
test: Update golden partial file

Updates the golden partial file to account for the newly added test

alxhub commented:
This PR was merged into the repository by commit 8d09e9e01380efd1f493139c9ae3eb4573e4e061.
